storage/jellyfish_merkle/src/node_type/mod.rs

```diff
@@ -0,0 +1,505 @@
+                child_half_start,
+                width,
+                (existence_bitmap, leaf_bitmap),
+            );
+
+            if range_existence_bitmap == 0 {
+                // No child in this range.
+                return (None, siblings);
+            } else if range_existence_bitmap.count_ones() == 1
+                && (range_leaf_bitmap.count_ones() == 1 || width == 1)
+            {
+                // Return the only 1 leaf child under this subtree or reach the lowest level
+                // Even this leaf child is not the n-th child, it should be returned instead of
+                // `None` because it's existence indirectly proves the n-th child doesn't exist.
+                // Please read proof format for details.
+                let only_child_index = range_existence_bitmap.trailing_zeros() as u8;
+                return (
+                    {
+                        let only_child_version = self
+                            .child(only_child_index)
+                            .unwrap_or_else(|| {
+                                panic!(
+                                    "Corrupted internal node: child_bitmap indicates \
+                                     the existence of a non-exist child at index {}",
+                                    only_child_index
+                                )
+                            })
+                            .version;
+                        Some(node_key.gen_child_node_key(only_child_version, only_child_index))
+                    },
+                    siblings,
+                );
+            }
+        }
+        unreachable!()
+    }
+}
+
+impl LeafNode {
+    /// Creates a new leaf node.
+    pub fn new(account_key: HashValue, blob: AccountStateBlob) -> Self {
+        let blob_hash = blob.hash();
+        Self {
+            account_key,
+            blob_hash,
+            blob,
+        }
+    }
+
+    /// Gets the account key, the hashed account address.
+    pub fn account_key(&self) -> HashValue {
+        self.account_key
+    }
+
+    /// Gets the hash of associated blob.
+    pub fn blob_hash(&self) -> HashValue {
+        self.blob_hash
+    }
+
+    /// Gets the associated blob itself.
+    pub fn blob(&self) -> &AccountStateBlob {
+        &self.blob
+    }
+}
+
+impl Node {
+    /// Creates the [`Internal`](Node::Internal) variant.
+    pub fn new_internal(children: HashMap<u8, Child>) -> Self {
+        Node::Internal(InternalNode::new(children))
+    }
+
+    /// Creates the [`Leaf`](Node::Leaf) variant.
+    pub fn new_leaf(account_key: HashValue, blob: AccountStateBlob) -> Self {
+        Node::Leaf(LeafNode::new(account_key, blob))
+    }
+
+    /// Returns `true` if the node is a leaf node.
+    pub fn is_leaf(&self) -> bool {
+        match self {
+            Node::Leaf(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Serializes to bytes for physical storage.
+    pub fn encode(&self) -> Result<Vec<u8>> {
```

Comment:
I don't really like this serde/manual-encoding mixed approach. But I guess this is temporary anyways.

Reply:
Yea. I will leave it for @zamsden